hail/src/main/scala/is/hail/io/fs/AzureStorageFS.scala

```diff
@@ -0,0 +1,280 @@
+    }
+  }
+}
+
+
+class AzureStorageFS(val credentialsJSON: Option[String] = None) extends FS {
+  @transient private lazy val serviceClientCache = credentialsJSON match {
+    case None =>
+      val credential: DefaultAzureCredential = new DefaultAzureCredentialBuilder().build()
+      new AzureBlobServiceClientCache(credential)
+    case Some(keyData) =>
+      val kvs = JsonMethods.parse(keyData) match {
+        case json4s.JObject(values) => values.toMap
+      }
+
+      val appId = kvs("appId").asInstanceOf[json4s.JString].s
+      val password = kvs("password").asInstanceOf[json4s.JString].s
+      val tenant = kvs("tenant").asInstanceOf[json4s.JString].s
+
+      val clientSecretCredential: ClientSecretCredential = new ClientSecretCredentialBuilder()
+        .clientId(appId)
+        .clientSecret(password)
+        .tenantId(tenant)
+        .build()
+      new AzureBlobServiceClientCache(clientSecretCredential)
+  }
+
+  def getBlobServiceClient(account: String): BlobServiceClient = {
+    serviceClientCache.getServiceClient(account)
+  }
+
+  def getBlobClient(account: String, container: String, path: String): BlobClient = {
+    getBlobServiceClient(account).getBlobContainerClient(container).getBlobClient(path)
+  }
+
+  def getContainerClient(account: String, container: String): BlobContainerClient = {
+    getBlobServiceClient(account).getBlobContainerClient(container)
+  }
+
+  def openNoCompression(filename: String): SeekableDataInputStream = {
+    val (account, container, path) = getAccountContainerPath(filename)
+    val blobClient: BlobClient = getBlobClient(account, container, path)
+    val blobProperties = blobClient.getProperties
+    val blobSize = blobProperties.getBlobSize
+
+    val is: SeekableInputStream = new FSSeekableInputStream {
+      private[this] val client: BlobClient = blobClient
+
+      def fill(): Unit = {
+        bb.clear()
+
+        val outputStreamToBuffer: OutputStream = (i: Int) => {
+          bb.put(i.toByte)
+        }
+
+        // calculate the minimum of the bytes remaining in the buffer and the bytes remaining
+        // to be read in the blob
+        val count = Math.min(blobSize - pos, bb.remaining())
+        if (count <= 0) {
+          eof = true
+          return
+        }
+
+        client.downloadStreamWithResponse(
+          outputStreamToBuffer, new BlobRange(pos, count),
+          null, null, false, Duration.ofMinutes(1), null)
+        pos += count
+        bb.flip()
+
+        assert(bb.position() == 0 && bb.remaining() > 0)
+      }
+
+      override def seek(newPos: Long): Unit = {
+        bb.clear()
+        bb.limit(0)
+        pos = newPos
+      }
+    }
+
+    new WrappedSeekableDataInputStream(is)
+  }
+
+  def createNoCompression(filename: String): PositionedDataOutputStream = {
+    val (account, container, path) = getAccountContainerPath(filename)
+    val appendClient: AppendBlobClient = getBlobClient(account, container, path).getAppendBlobClient
```

Comment:
This type isn't required, right? AFAIK, we generally avoid type annotations in Scala if they're unnecessary.